sql/src/main/java/io/crate/lucene/GenericFunctionQuery.java

```diff
@@ -0,0 +1,236 @@
+        this.function = function;
+        // inner loop iterates over expressions - call toArray to avoid iterator allocations
+        this.expressions = expressions.toArray(new LuceneCollectorExpression[0]);
+        this.collectorContext = collectorContext;
+        this.condition = condition;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        if (!super.equals(o)) return false;
+
+        GenericFunctionQuery that = (GenericFunctionQuery) o;
+
+        return function.equals(that.function);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = super.hashCode();
+        result = 31 * result + function.hashCode();
+        return result;
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, boolean needsScores) throws IOException {
+        return new Weight(this) {
+            @Override
+            public void extractTerms(Set<Term> terms) {
+            }
+
+            @Override
+            public Explanation explain(LeafReaderContext context, int doc) throws IOException {
+                final Scorer s = scorer(context);
+                final boolean match;
+                if (s == null) {
+                    match = false;
+                } else {
+                    final TwoPhaseIterator twoPhase = s.twoPhaseIterator();
+                    if (twoPhase == null) {
+                        match = s.iterator().advance(doc) == doc;
+                    } else {
+                        match = twoPhase.approximation().advance(doc) == doc && twoPhase.matches();
+                    }
+                }
+                if (match) {
+                    assert s.score() == 0f;
+                    return Explanation.match(0f, "Match on id " + doc);
+                } else {
+                    return Explanation.match(0f, "No match on id " + doc);
+                }
+            }
+
+            @Override
+            public float getValueForNormalization() throws IOException {
+                return 0;
+            }
+
+            @Override
+            public void normalize(float norm, float boost) {
+            }
+
+            @Override
+            public Scorer scorer(LeafReaderContext context) throws IOException {
+                final DocIdSet set = getDocIdSet(context);
+                if (set == null) {
+                    return null;
+                }
+                // lucene 5.5.0 FilteredQuery switches to random-access logic if bits are present.
+                // this logic isn't part of this GenericFunctionQuery because FilteredDocIdSet never has bits
+                // this assertion is a safety measure - if bits() suddenly is no longer null this code here should be updated
+                assert set.bits() == null : "bits are never set because getDocIdSet returns FilteredDocIdSet";
```

Comment:
hm can you reword the message? something like "... should never.." or ".. are expected to be never.. "? would make it more clear if it happens..